File: src/data/codes/kmodes.m
clc;  % k-modes
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'kmodes_student_profiles.csv';
K = 3;
maxIter = 20;

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    Gender = categorical(["Male";"Female";"Male";"Female";"Male";"Female";"Male";"Female";"Male";"Female"]);
    Branch = categorical(["CSE";"IT";"ECE";"CSE";"IT";"ECE";"CSE";"IT";"ECE";"CSE"]);
    Residence = categorical(["Hostel";"Day";"Hostel";"Day";"Hostel";"Hostel";"Day";"Day";"Hostel";"Day"]);
    Study_Style = categorical(["Night";"Morning";"Night";"Morning";"Night";"Night";"Morning";"Morning";"Night";"Morning"]);
    Attendance_Level = categorical(["High";"Medium";"Low";"High";"Medium";"Low";"High";"Medium";"Low";"High"]);
    Result_Status = categorical(["Pass";"Pass";"Fail";"Pass";"Pass";"Fail";"Pass";"Pass";"Fail";"Pass"]);

    data = table(Gender, Branch, Residence, Study_Style, Attendance_Level, Result_Status);
else
    data = readtable(csvFile);
    % Expected all categorical columns
end

disp(data);

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = data{:,:};   % categorical table to cell-like content
[n, m] = size(X);

%% Convert to string matrix for easy comparison
Xstr = strings(n,m);
for i = 1:n
    for j = 1:m
        Xstr(i,j) = string(X{i,j});
    end
end

%% =========================
% INITIALIZE MODES
% ==========================
rng(14);
randIdx = randperm(n, K);
modes = Xstr(randIdx, :);

clusterIdx = zeros(n,1);

%% =========================
% K-MODES ITERATION
% ==========================
for iter = 1:maxIter
    oldClusterIdx = clusterIdx;

    % Assignment step
    for i = 1:n
        dist = zeros(K,1);
        for k = 1:K
            dist(k) = sum(Xstr(i,:) ~= modes(k,:));   % simple matching distance
        end
        [~, clusterIdx(i)] = min(dist);
    end

    % Update step
    for k = 1:K
        clusterPoints = Xstr(clusterIdx == k, :);
        if ~isempty(clusterPoints)
            for j = 1:m
                vals = unique(clusterPoints(:,j));
                counts = zeros(length(vals),1);
                for v = 1:length(vals)
                    counts(v) = sum(clusterPoints(:,j) == vals(v));
                end
                [~, maxIdx] = max(counts);
                modes(k,j) = vals(maxIdx);
            end
        end
    end

    if isequal(oldClusterIdx, clusterIdx)
        break;
    end
end

%% =========================
% TOTAL COST
% ==========================
totalCost = 0;
for i = 1:n
    totalCost = totalCost + sum(Xstr(i,:) ~= modes(clusterIdx(i),:));
end

fprintf('\nTotal clustering cost = %d\n', totalCost);

disp('Cluster assignments:');
disp(clusterIdx);

disp('Modes of clusters:');
disp(modes);

File: src/data/codes/kmodes.py
"""K-modes (categorical clustering) — manual matching distance
# pip install numpy pandas
"""
import numpy as np
import pandas as pd

USE_SYNTHETIC = True
K = 3
MAX_ITER = 20
RNG = np.random.default_rng(14)

if USE_SYNTHETIC:
    data = pd.DataFrame(
        {
            "Gender": ["Male", "Female"] * 5,
            "Branch": ["CSE", "IT", "ECE"] * 3 + ["CSE", "IT"],
            "Residence": ["Hostel", "Day"] * 5,
            "Study_Style": ["Night", "Morning"] * 5,
            "Attendance_Level": ["High", "Medium", "Low"] * 3 + ["High", "Medium"],
            "Result_Status": ["Pass", "Pass", "Fail", "Pass", "Pass", "Fail", "Pass", "Pass", "Fail", "Pass"],
        }
    )
else:
    data = pd.read_csv("kmodes_student_profiles.csv")

print(data)
data = data.dropna().drop_duplicates()
Xstr = data.astype(str).values
n, m = Xstr.shape

rand_idx = RNG.permutation(n)[:K]
modes = Xstr[rand_idx].copy()
cluster_idx = np.zeros(n, dtype=int)

for _iter in range(MAX_ITER):
    old = cluster_idx.copy()
    for i in range(n):
        dist = np.zeros(K)
        for k in range(K):
            dist[k] = np.sum(Xstr[i] != modes[k])
        cluster_idx[i] = int(np.argmin(dist))
    for k in range(K):
        mask = cluster_idx == k
        if not np.any(mask):
            continue
        cluster_pts = Xstr[mask]
        for j in range(m):
            vals, counts = np.unique(cluster_pts[:, j], return_counts=True)
            modes[k, j] = vals[np.argmax(counts)]
    if np.array_equal(old, cluster_idx):
        break

cost = sum(np.sum(Xstr[i] != modes[cluster_idx[i]]) for i in range(n))
print(f"\nTotal clustering cost = {cost}")
print("Cluster assignments:", cluster_idx)
print("Modes:\n", modes)
